File: 5.Prefix Sums - MinAvgTwoSlice/answer.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")


def solution(A):
    # write your code in Python 3.6
    currentMin = 100000
    ans = 0
    for i in range(len(A)-1):
        # Check two slices
        currentSum = A[i] + A[i+1]
        avgSum = (currentSum / 2)
        if(avgSum < currentMin):
            currentMin = avgSum
            ans = i
            
        # Check three slices (It is possible for three slices to have lower average than two slice). There is no need to check further slices as it can be composed of the two and three slices.
        if(i + 2 < len(A)):
            avgSum = ((currentSum + A[i+2]) / 3)
            if(avgSum < currentMin):
                ans = i
                currentMin = avgSum

    return ans

File: 5.Prefix Sums - GenomicRangeQuery/answer.js
// you can write to stdout for debugging purposes, e.g.
// console.log('this is a debug message');

function solution(S, P, Q) {
    // write your code in JavaScript (Node.js 8.9.4)
    let answer = []
    let nucleotides = {
        "A" : 0,
        "C" : 0,
        "G" : 0,
        "T" : 0
    }
    let nucleoPrefixSum = []
    for(const char of S){
        nucleoPrefixSum.push(Object.assign({}, nucleotides))
        nucleotides[char] += 1
    }
    
    let arrSize = P.length
    for (let i =0; i < arrSize; i++){
        if(nucleoPrefixSum[Q[i]]["A"] - nucleoPrefixSum[P[i]]["A"] > 0)
            answer.push(1)
        else if(nucleoPrefixSum[Q[i]]["C"] - nucleoPrefixSum[P[i]]["C"] > 0)
            answer.push(2)
        else if(nucleoPrefixSum[Q[i]]["G"] - nucleoPrefixSum[P[i]]["G"] > 0)
            answer.push(3)
        else
            answer.push(4)
    }
    
    return answer
}
